pkg/cloud/services/eks/roles.go

```diff
@@ -0,0 +1,362 @@
+/*
```

Comment:
Generally unsure of doing a bunch of IAM work inside the controller as it then has to be super-privileged over AWS accounts, vs the current model where clusterawsadm is run administratively as a one-time operation and then the controller can be run with a much smaller set of permissions.

Reply:
I know what you mean. The reason i added this was that with EKS its not necessarily a one-time operation and that it could be per cluster (but the roles could be reused as well)

Reply:
Still thinking about this one.  This code is creating the cluster service iam role: https://docs.aws.amazon.com/eks/latest/userguide/service_IAM_role.html.

Previously i have always created a role per cluster (and this is also what eksctl does) as this allows the role to be tailored to the role of the cluster. For example by adding extra policies:
https://github.com/richardcase/cluster-api-provider-aws/blob/ekscontrolplane/exp/api/v1alpha3/awsmanagedcontrolplane_types.go#L62

Like you say we could try and put it into `clusterawsadm` and perhaps create a default role when creating the initial stack. And then have a subcommand where a new role could be created with additional policies if a cluster needed something unique. Although the role then is managed via a separate mechanism and isn't attached to the cluster reconciliation.

If we keep it in the controller then the changes to the permissions can be scoped to this:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1780#issuecomment-667997298

@randomvariable @rudoi @detiber - happy to go with what you think is best here.

Reply:
You can specify a role in the Spec right?

I'd be happy to leave this as-is for now and then add something optional to `clusterawsadm` in a follow-up.

Reply:
The difference is that `eksctl` is running on the command line as the end user. In a more permanent management cluster, there's a privilege escalation issues that Kubernetes RBAC can't protect against.

In the EC2 based controller, we've been pretty strict in that the default cloudformation template only allows instance profiles based on IAM roles using the `cluster-api-provider-aws.sigs.k8s.io` suffix, allowing AWS admins to restrict what roles EC2 instances can be created with.

In particular, allowing `RoleAdditionalPolicies` means anyone who can create a cluster can create a role with administrative access.

Would it be possible to have this behaviour opt-in as a flag?

Reply:
Thats a good point about it being a different use case.

We could have `RoleAdditionalPolicies` as an opt in feature flag, no problem at all. If we did that would you be more comfortable having the IAM role creation in the controller?

Reply:
Can we have at least the ability to disable write access to IAM entirely? Can also optionally include one for RoleAdditionalPolicies

Reply:
Sounds like a plan

Reply:
I'm going to add 2 additional feature flags to cover this.